cmd/snap/inhibit.go

```diff
@@ -34,143 +34,160 @@ import (
 	"github.com/snapcore/snapd/usersession/client"
 )
 
-func waitWhileInhibited(snapName string) error {
-	hint, _, err := runinhibit.IsLocked(snapName)
-	if err != nil {
-		return err
-	}
-	if hint == runinhibit.HintNotInhibited {
-		return nil
+var runinhibitWaitWhileInhibited = runinhibit.WaitWhileInhibited
+
+func waitWhileInhibited(snapName string) (err error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer func() {
+		if err != nil {
+			cancel()
+		}
+	}()
+
+	flow := newInhibitionFlow(snapName)
+	notified := false
+	inhibited := func(hint runinhibit.Hint, inhibitInfo *runinhibit.InhibitInfo) (cont bool, err error) {
+		if !notified {
+			// wait for HintInhibitedForRefresh set by gate-auto-refresh hook handler
+			// when it has finished; the hook starts with HintInhibitedGateRefresh lock
+			// and then either unlocks it or changes to HintInhibitedForRefresh (see
+			// gateAutoRefreshHookHandler in hooks.go).
+			if hint != runinhibit.HintInhibitedForRefresh {
+				return false, nil
+			}
+			// Start notification flow.
+			if err := flow.StartInhibitionNotification(ctx); err != nil {
+				return true, err
+			}
+			// Make sure we call notification flow only once.
+			notified = true
+		}
+		return false, nil
 	}
 
-	// wait for HintInhibitedForRefresh set by gate-auto-refresh hook handler
-	// when it has finished; the hook starts with HintInhibitedGateRefresh lock
-	// and then either unlocks it or changes to HintInhibitedForRefresh (see
-	// gateAutoRefreshHookHandler in hooks.go).
-	// waitInhibitUnlock will return also on HintNotInhibited.
-	notInhibited, err := waitInhibitUnlock(snapName, runinhibit.HintInhibitedForRefresh)
+	hintFlock, err := runinhibitWaitWhileInhibited(snapName, nil, inhibited, 500*time.Millisecond)
```

Comment:
would it make to pass a function for notInhibited argument which would call flow.FinishInhibitionNotification() ? AFAICT WaitWhileInhibited seems to DTRT inside.

Reply:
it will work, but is there a reason to do it like this?

Reply:
Otherwise the notInhibited argument of runinhibit.WaitWhileInhibited doesn't appear to be used anywhere, and if so perhaps it's not needed in that call. Then if it already does the right thing, there is no need to repeat the behavior here.